src/Evolution/Systems/NewtonianEuler/PrimitiveVariables.cpp

```diff
@@ -0,0 +1,65 @@
+// Distributed under the MIT License.
+// See LICENSE.txt for details.
+
+#include "Evolution/Systems/NewtonianEuler/PrimitiveVariables.hpp"
+
+#include "DataStructures/DataVector.hpp"
+#include "DataStructures/Tensor/Tensor.hpp"
+#include "Utilities/GenerateInstantiations.hpp"
+
+namespace NewtonianEuler {
+template <size_t Dim, typename DataType>
+tnsr::I<DataType, Dim> compute_velocity_from_conserved_quantities(
+    const Scalar<DataType>& density, const tnsr::I<DataType, Dim>& momentum,
+    const Scalar<DataType>& /*energy*/) noexcept {
+  auto velocity = make_with_value<tnsr::I<DataType, Dim>>(density, 0.0);
+  auto one_over_density = make_with_value<Scalar<DataType>>(density, 0.0);
+
+  get(one_over_density) = 1.0 / get(density);
+  for (size_t i = 0; i < Dim; ++i) {
+    velocity.get(i) = momentum.get(i) * one_over_density.get();
+  }
+  return velocity;
+}
+
+template <size_t Dim, typename DataType>
+Scalar<DataType> compute_specific_internal_energy_from_conserved_quantities(
+    const Scalar<DataType>& density, const tnsr::I<DataType, Dim>& momentum,
+    const Scalar<DataType>& energy) noexcept {
+  auto specific_internal_energy =
+      make_with_value<Scalar<DataType>>(density, 0.0);
+  auto one_over_density = make_with_value<Scalar<DataType>>(density, 0.0);
+  const auto velocity = compute_velocity_from_conserved_quantities<Dim>(
+      density, momentum, energy);
+  auto velocity_squared = make_with_value<Scalar<DataType>>(density, 0.0);
+
+  get(one_over_density) = 1.0 / get(density);
+  for (size_t i = 0; i < Dim; ++i) {
```

Comment:
just add a `magnitude_squared` function to Tensor/EagerMath/Magnitude.hpp and have this and the `magnitude` functions call that.